docs/content/0.1.0/docs/developers/functions.rst

```diff
@@ -0,0 +1,204 @@
+
+2. Translation to SQLAlchemy. In order to run queries, DJ parses the metric definitions (written in ANSI SQL), and converts them to a SQLAlchemy query object, so it can be translated to different dialects (Hive, Trino, Postgres, etc.). Some functions are easier to translate, specially if they are already defined in ``sqlalchemy.sql.functions``. Others, like ``DATE_TRUNC``, are more complex because they are dialect specific.
+
+Supported functions
+-------------------
+
+``AVG``
+-------
+
+Return the average of a given column:
+
+.. code-block:: sql
+
+    > SELECT AVG(column);
+    1.2
+
+``COALESCE``
+------------
+
+Return the first non-null value:
+
+.. code-block:: sql
+
+    > SELECT column_a, column_b FROM some_table;
+    1, NULL
+    NULL, 10
+    NULL, NULL
+    > SELECT COALESCE(column_a, column_b, -1) FROM some_table;
+    1
+    10
+    -1
+
+``COUNT``
+---------
+
+The humble ``COUNT()``.
+
+.. code-block:: sql
+
+    > SELECT COUNT(*) FROM some_table;
+    10
+    > SELECT COUNT(1) FROM some_table;
+    10
+    > SELECT COUNT(column) FROM some_table;  -- ignores NULLs
+    5
+
+``DATE_TRUNC``
+--------------
+
+Truncate a ``DATETIME`` column to a given resolution:
+
+.. code-block:: sql
+
+    > SELECT DATE_TRUNC('minute', CAST('2022-01-01T12:34:56Z' AS TIMESTAMP);
+    2022-01-01T12:34:00Z
+
+``MAX``
+-------
+
+Return the maximum value from a column:
+
+.. code-block:: sql
+
+    > SELECT MAX(column);
+
+``MIN``
+-------
+
+Return the minimum value from a column:
+
+.. code-block:: sql
+
+    > SELECT MIN(column);
+    1
+
+``SUM``
+-------
+
+Return the sum of a given column:
+
+.. code-block:: sql
+
+    > SELECT SUM(sales)
+    12345
+
+Adding new functions
+--------------------
+
+Let's look at the ``COUNT`` function in DJ:
+
+.. code-block:: python
+
+    from sqlalchemy.sql import func
+    from sqlalchemy.sql.schema import Column as SqlaColumn
+
+    from dj.models.column import Column
+    from dj.typing import ColumnType
+
+
+    class Count(Function):
+        """
+        The ``COUNT`` function.
+        """
+
+        is_aggregation = True
+
+        @staticmethod
+        def infer_type(argument: Union[Column, "Wildcard", int]) -> ColumnType:
+            return ColumnType.INT
+
+        @staticmethod
+        def get_sqla_function(
+            argument: Union[SqlaColumn, str, int],
+            *,
+            dialect: Optional[str] = None,
+        ) -> SqlaFunction:
+            return func.count(argument)
+
+
+The first method, ``infer_type``, is responsible for type inference. The function is usually called as ``COUNT(column)``, ``COUNT(1)`` or ``COUNT(*)``, so we define the input argument as either a column, a star, or a number. In retrospect we could have also added a default value, to make ``COUNT`` valid. We can see that the method always return an integer.
```

Comment:
This will probably be changing relatively soon right? I imagine that `infer_type` should expect something like `*args: ColumnType` instead

Reply:
Ah thanks for reviewing this file @CircArgs, it was one of the files that was already in the docs directory and I included it un-edited just to preserve it (moving it to another folder caused it to show up in this PR's diff). You're right that much of this is going to change soon.